### PR TITLE
Graceful shutdown

### PR DESCRIPTION
**Pull request by pjdufour-truss**
## Description

This PR adds support for a graceful shutdown of the server.  This has 2 benefits.

First, this PR will decrease our risk of data corruption during deployments.  The server will now wait for all client connections to close and will then immediately shut down.  Currently the server waits 30 seconds and then quits regardless if there are any open connections.  This assumes any database transactions are complete within the HTTP request/response cycle.

Second, this PR will speed up our deployments since each container won't wait 30 seconds to begin shutdown, but will start to shutdown ASAP (assuming no open connections).

For context see:

https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_StopTask.html

## Reviewer Notes

TBD: I'll add some notes on how to test thoroughly.

## Setup

None

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](./docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](./docs/backend.md#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](./docs/database.md#zero-downtime-migrations))
  * [ ] Have been communicated to #dp3-engineering
  * [ ] Secure migrations have been tested using `bin/run-prod-migrations`
* [ ] There are no aXe warnings for UI.
* [ ] This works in IE.
* Any new client dependencies (Google Analytics, hosted libraries, CDNs, etc) have been:
  * [ ] Communicated to @ntwyman
  * [ ] Added to the list of [network dependencies](https://github.com/transcom/mymove#client-network-dependencies)
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/160428223) for this change

## Screenshots

None

### CODE REVIEW

File: cmd/webserver/main.go [hunk 1]
```diff
@@ -171,6 +175,7 @@ func initFlags(flag *pflag.FlagSet) {
 	flag.String("env", "development", "The environment to run in, which configures the database.")
 	flag.String("interface", "", "The interface spec to listen for connections on. Default is all.")
 	flag.String("service-name", "app", "The service name identifies the application for instrumentation.")
+	flag.Duration("graceful-shutdown-timeout", 25*time.Second, "The duration for which the server gracefully wait for existing connections to finish")
```
Comment (chrisgilmerproj) [inline cmd/webserver/main.go]: We should add a comment that ECS only gives us 30 seconds, so any number less than 30 should be used.

File: cmd/webserver/main.go [hunk 2]
```diff
@@ -1188,11 +1209,64 @@ func main() {
...
 	}()
 
-	logger.Fatal("listener error", zap.Error(<-errChan))
+	wg.Wait()
+	logger.Info("All listeners are shutdown")
+	logger.Sync()
+
+	shutdownError := false
+	shutdownErrors.Range(func(key, value interface{}) bool {
+		srv := key.(*server.NamedServer)
```
Comment (chrisgilmerproj) [inline cmd/webserver/main.go]: Should we catch errors here just to be safe? `srv, ok := key.(*server.NamedServer)`?

File: cmd/webserver/main.go [hunk 3]
```diff
@@ -1188,11 +1209,64 @@ func main() {
 	if err != nil {
 		logger.Fatal("error creating mutual-tls server", zap.Error(err))
 	}
+	go startListener(mutualTLSServer, logger, true)
+
+	quit := make(chan os.Signal, 1)
+
+	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
+
+	sig := <-quit
```
Comment (chrisgilmerproj) [inline cmd/webserver/main.go]: I'm still pretty new to channels. Does this essentially stop the code here and wait for the output from the quit channel?

File: cmd/webserver/main.go (hunk 2 above)
Comment (pjdufour-truss) [reply cmd/webserver/main.go]: Sure, why not.  :)

File: cmd/webserver/main.go (hunk 1 above)
Comment (pjdufour-truss) [reply cmd/webserver/main.go]: Added to usage

File: cmd/webserver/main.go (hunk 3 above)
Comment (pjdufour-truss) [reply cmd/webserver/main.go]: Added comments to code.

**Pull request merged**